File: pipeline/bin/merge_pcs.R
#!/usr/bin/env Rscript
library(optparse)
library(data.table)
library(ggplot2)

option_list = list(
  make_option(c("--cov_file"), type="character"),
  make_option(c("--pcs_file"), type="character"),
  make_option(c("--eval_file"), type="character"),
  make_option(c("--outdir"), type="character")
)

opt_parser = OptionParser(option_list=option_list)
opt = parse_args(opt_parser)

dir.create(opt$outdir, showWarnings = FALSE, recursive = TRUE)

# 1. Load data
covs <- fread(opt$cov_file)
pcs <- fread(opt$pcs_file)

# Format PC file (plink2 pca.eigenvec often uses '#FID' as first header)
if ("#FID" %in% colnames(pcs) && !"FID" %in% colnames(pcs)) {
  setnames(pcs, "#FID", "FID")
}

# If headerless, fall back and create names
if (!all(c("FID", "IID") %in% colnames(pcs))) {
  pcs <- fread(opt$pcs_file, header = FALSE)
  if (ncol(pcs) < 3) {
    stop("PC file has too few columns (expected FID IID PC1 ...).")
  }
  setnames(pcs, c("FID", "IID", paste0("PC", 1:(ncol(pcs) - 2))))
}

covs[, FID := as.character(FID)]
covs[, IID := as.character(IID)]
pcs[, FID := as.character(FID)]
pcs[, IID := as.character(IID)]

# 2. Merge
merged <- merge(covs, pcs, by = c("FID", "IID"), all.x = TRUE, sort = FALSE)

# Ensure covariate column names are safe for downstream tools (keep FID/IID intact)
id_cols <- c("FID", "IID")
other_cols <- setdiff(colnames(merged), id_cols)
if (length(other_cols) > 0) {
  safe_other <- make.names(other_cols, unique = TRUE)
  setnames(merged, old = other_cols, new = safe_other)
}

out_file <- file.path(opt$outdir, "covariates.cov")
fwrite(merged, out_file, sep = "\t", quote = FALSE, na = "NA")

# 3. Scree Plot
if (file.exists(opt$eval_file)) {
  evals <- fread(opt$eval_file, header = FALSE)
  colnames(evals) <- "Val"
  evals[, PC := seq_len(.N)]
  evals[, PVE := (Val / sum(Val)) * 100]
  
  p <- ggplot(evals, aes(x = factor(PC), y = PVE)) +
    geom_bar(stat = "identity", fill = "steelblue") +
    geom_text(aes(label = sprintf("%.1f%%", PVE)), vjust = -0.5, size = 3) +
    theme_minimal() +
    labs(title = "Scree Plot",
         x = "Principal Component",
         y = "Proportion of Variance Explained (%)") +
    scale_y_continuous(expand = expansion(mult = c(0, 0.1)))

  ggsave(file.path(opt$outdir, "scree_plot.png"), p, width = 8, height = 5, dpi = 300)
}

File: pipeline/bin/merge_pcs.py
#!/usr/bin/env python3

from __future__ import annotations

import argparse
from pathlib import Path

import pandas as pd

from config import ensure_string_ids, make_safe_names, read_table_auto


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser()
    parser.add_argument("--cov_file", required=True)
    parser.add_argument("--pcs_file", required=True)
    parser.add_argument("--eval_file", required=True)
    parser.add_argument("--outdir", required=True)
    return parser.parse_args()


def load_pcs_table(pcs_file: str) -> pd.DataFrame:
    pcs = read_table_auto(pcs_file)
    if "#FID" in pcs.columns and "FID" not in pcs.columns:
        pcs = pcs.rename(columns={"#FID": "FID"})

    if {"FID", "IID"}.issubset(pcs.columns):
        return pcs

    pcs = pd.read_csv(pcs_file, sep=r"\s+", engine="python", header=None)
    if pcs.shape[1] < 3:
        raise ValueError("PC file has too few columns (expected FID IID PC1 ...).")

    pcs.columns = ["FID", "IID"] + [f"PC{i}" for i in range(1, pcs.shape[1] - 1)]
    return pcs


def main() -> None:
    args = parse_args()
    outdir = Path(args.outdir)
    outdir.mkdir(parents=True, exist_ok=True)

    covariates = read_table_auto(args.cov_file)
    pcs = load_pcs_table(args.pcs_file)

    covariates = ensure_string_ids(covariates)
    pcs = ensure_string_ids(pcs)

    merged = covariates.merge(pcs, on=["FID", "IID"], how="left", sort=False)

    other_columns = [col for col in merged.columns if col not in {"FID", "IID"}]
    if other_columns:
        merged = merged.rename(columns=dict(zip(other_columns, make_safe_names(other_columns))))

    merged.to_csv(outdir / "covariates.cov", sep="\t", index=False, na_rep="NA")

if __name__ == "__main__":
    main()
